be/src/storage/rowset/rowset_writer.cpp

```diff
@@ -209,104 +211,161 @@ StatusOr<RowsetSharedPtr> RowsetWriter::build() {
+        partial_rowset_footer->set_size(segment_pb.partial_footer_size());
+    }
 
-        VLOG(2) << "Flush segment to " << path << " size " << segment_pb.data_size();
+    // 3. update statistic
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        _total_data_size += segment_pb.data_size();
+        _total_index_size += segment_pb.index_size();
+        _num_rows_written += segment_pb.num_rows();
+        _total_row_size += segment_pb.row_size();
+        _num_segment++;
     }
 
-    if (segment_pb.has_delete_path()) {
-        // 1. create delete file
-        auto path =
-                Rowset::segment_del_file_path(_context.rowset_path_prefix, _context.rowset_id, segment_pb.delete_id());
-        // use MUST_CREATE make sure atomic
-        ASSIGN_OR_RETURN(auto wfile, _fs->new_writable_file(path));
-
-        // 2. flush delete file
-        auto writer = std::make_unique<SegmentFileWriter>(wfile.get());
-
-        butil::IOBuf delete_data;
-        int64_t remaining_bytes = data.cutn(&delete_data, segment_pb.delete_data_size());
-        if (remaining_bytes != segment_pb.delete_data_size()) {
-            return Status::InternalError(fmt::format("segment {} delete size {} not equal attachment size {}",
-                                                     segment_pb.DebugString(), remaining_bytes,
-                                                     segment_pb.delete_data_size()));
-        }
-        while (remaining_bytes > 0) {
-            auto written_bytes = delete_data.cut_into_writer(writer.get(), remaining_bytes);
-            if (written_bytes < 0) {
-                return io::io_error(wfile->filename(), errno);
-            }
-            remaining_bytes -= written_bytes;
-        }
-        if (remaining_bytes != 0) {
-            return Status::InternalError(fmt::format("segment delete file {} write size {} not equal expected size {}",
-                                                     wfile->filename(), segment_pb.delete_data_size() - remaining_bytes,
-                                                     segment_pb.delete_data_size()));
+    VLOG(2) << "Flush segment to " << path << " size " << segment_pb.data_size();
+
+    return Status::OK();
+}
+
+Status RowsetWriter::_flush_delete_file(const SegmentPB& segment_pb, butil::IOBuf& data) {
+    // 1. create delete file
+    auto path = Rowset::segment_del_file_path(_context.rowset_path_prefix, _context.rowset_id, segment_pb.delete_id());
+    // use MUST_CREATE make sure atomic
+    ASSIGN_OR_RETURN(auto wfile, _fs->new_writable_file(path));
+
+    // 2. flush delete file
+    auto writer = std::make_unique<SegmentFileWriter>(wfile.get());
+
+    butil::IOBuf delete_data;
+    int64_t remaining_bytes = data.cutn(&delete_data, segment_pb.delete_data_size());
+    if (remaining_bytes != segment_pb.delete_data_size()) {
+        return Status::InternalError(fmt::format("segment {} delete size {} not equal attachment size {}",
```

Comment:
How about the segment_pb size? Maybe it is too large to place the debug string in the returned status?

Reply:
change to segment's filename.